File: c/math/pascal.c
/* Print 10 rows of Pascal's triangle, each row rewritten in place from its right hand end. */
/* Build: cc -o pascal pascal.c */

#include <stdio.h>

#define ROWS 10

int main(void)
{
    int row[ROWS + 1] = {0};
    int length;

    row[0] = 1;

    for (length = 1; length <= ROWS; length++) {
        int i;

        for (i = 0; i < length; i++) {
            printf("%s%d", i > 0 ? " " : "", row[i]);
        }
        putchar('\n');

        for (i = length; i > 0; i--) {
            row[i] += row[i - 1];
        }
    }
    return 0;
}

File: zig/math/pascal.zig
// Print 10 rows of Pascal's triangle, each row rewritten in place from its right hand end. (written against Zig 0.16)
// Build: zig build-exe pascal.zig

const std = @import("std");

const rows = 10;

pub fn main() void {
    var row = [_]u64{0} ** (rows + 1);
    row[0] = 1;

    var length: usize = 1;
    while (length <= rows) : (length += 1) {
        for (row[0..length], 0..) |value, index| {
            if (index > 0) std.debug.print(" ", .{});
            std.debug.print("{d}", .{value});
        }
        std.debug.print("\n", .{});

        var i = length;
        while (i > 0) : (i -= 1) {
            row[i] += row[i - 1];
        }
    }
}
